### PR TITLE
storage: try to select an image ID when writing

### PR DESCRIPTION
**Pull request by nalind**
When saving a blob for an image, if an ID for the image wasn't specified, try to use the hex portion of the digest of the image's configuration blob as an ID.  In order to read that, we implement part of types.ImageSource on destination images.

### CODE REVIEW

File: storage/storage_image.go [hunk 1]
```diff
@@ -330,6 +366,16 @@ func (s *storageImageDestination) Commit() error {
 			lastLayer = layerList[len(layerList)-1]
 		}
 	}
+	if s.ID == "" {
```
Comment (runcom) [inline storage/storage_image.go]: Why can't we always use the config blob digest? It should be present for every image afaict and will allow us to group tags by id

Comment (nalind) [reply storage/storage_image.go]: schema1 images, like `kubernetes/pause`, don't have config blobs.

Comment (runcom) [reply storage/storage_image.go]: Mmm how is docker calculating the correct content addressable image ID for schema 1 images? We could copy that

Comment (mtrmac) [reply storage/storage_image.go]: Docker converts the schema1 manifest into a config, and uses the digest of that as an ID.

I think the `c/i/image` conversion is very similar, but it’s not tested to generate exactly the same IDs, and I don’t even know whether there is any expectation of this conversion to be stable and frozen forever; I’d be a bit surprised if it were.

**Comment by mtrmac** [comment]
Couldn’t this, like #305, instead built around the `manifest-editing`-like `c/i/manifest` subpackage, and just parsing the manifest, instead of involving all of this machinery?

**Comment by nalind** [comment]
@mtrmac Yup, this is a stab at providing something for https://github.com/kubernetes-incubator/cri-o/pull/654#issuecomment-338254868, but it would be replaced by logic in #305.  If we're going to get the same IDs for schema 1 images that dockerd computes, we'll also need diffIDs, so this isn't really going to work.  Closing.

**Pull request closed without merging**